### PR TITLE
improve obs-v4l2sink

### PR DESCRIPTION
**Pull request by notpeelz**
###### Motivation for this change

I started working on my own derivation of obs-v42lsink and realized that it had been added to nixpkgs 4 days ago. So I compared the two derivations and noticed a couple things that were done differently.

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

@grahamc @colemickens

### CODE REVIEW

File: pkgs/applications/video/obs-studio/v4l2sink.nix [hunk 1]
```diff
@@ -1,42 +1,49 @@
...
-    rev = "1ec3c8ada0e1040d867ce567f177be55cd278378";
-    sha256 = "03ah91cm1qz26k90mfx51l0d598i9bcmw39lkikjs1msm4c9dfxx";
+    rev = version;
+    sha256 = "0l4lavaywih5lzwgxcbnvdrxhpvkrmh56li06s3aryikngxwsk3z";
   };
 
-  nativeBuildInputs = [ cmake pkgconfig wrapQtAppsHook ];
-  buildInputs = [ obs-studio ];
+  nativeBuildInputs = [ cmake ];
+  buildInputs = [ qt5.qtbase.dev obs-studio ];
```
Comment (LouisTakePILLz) [inline pkgs/applications/video/obs-studio/v4l2sink.nix:21]: I'm not sure whether `wrapQtAppsHook` is needed (or preferable) in this case, considering it's a library.
I tried to follow the obs-v4l2sink README's build steps as closely as possible and this approach seemed to work.

Comment (LouisTakePILLz) [inline pkgs/applications/video/obs-studio/v4l2sink.nix:21]: I'm not sure whether `wrapQtAppsHook` is needed (or preferable) in this case, considering it's a library.
I tried to follow the obs-v4l2sink README's build steps as closely as possible and this approach seemed to work.

**Comment by LouisTakePILLz** [comment]
Related PR: #84153

@grahamc @colemickens

File: pkgs/applications/video/obs-studio/v4l2sink.nix (hunk 1 above)
Comment (grahamc) [reply pkgs/applications/video/obs-studio/v4l2sink.nix:21]: cc @jtojnar

**Comment by cole-h** [comment]
@ofborg eval

GitHub seems to be recover(ing/ed).

**Comment by LouisTakePILLz** [comment]
The 32-bit derivation can be tested with this:
```nix
{ pkgs ? import ./nixpkgs {}, is32bit ? false }:

with pkgs.lib;
let
  deps = pkgs: {
    inherit (pkgs)
      obs-studio
      obs-v4l2sink;
  };
  depsAttrset = if is32bit
    then deps pkgs.pkgsi686Linux.pkgs
    else deps pkgs;
in pkgs.stdenv.mkDerivation {
  name = "env";
  buildInputs = attrValues depsAttrset;
  shellHook = ''
    echo "obs-studio: ${depsAttrset.obs-studio.outPath}"
    echo "obs-v4l2sink: ${depsAttrset.obs-v4l2sink.outPath}"
    ln -sfT ${depsAttrset.obs-v4l2sink} ~/.config/obs-studio/plugins/v4l2sink
  '';
  unpackPhase = ":";
  installPhase = "touch $out";
}
```

```
$ nix-shell --arg is32bit true
```

File: pkgs/applications/video/obs-studio/v4l2sink.nix (hunk 1 above)
Comment (jtojnar) [reply pkgs/applications/video/obs-studio/v4l2sink.nix:21]: `wrapQtAppsHook` would be a no-op if there are no executables, AFAIK.

Comment (LouisTakePILLz) [reply pkgs/applications/video/obs-studio/v4l2sink.nix:21]: The original derivation builds just fine, so `wrapQtAppsHook` is at least supplying the qt5 libs (I've tested without `libsForQt5.callPackage` in `top-level/all-packages.nix`)

What I'm wondering is whether it would be better to drop the explicit `qt5.qtbase` dependency in favor of `wrapQtAppsHook`.

File: pkgs/applications/video/obs-studio/v4l2sink.nix [hunk 2]
```diff
@@ -1,42 +1,49 @@
...
+, fetchFromGitHub
+, cmake
+, qt5
+, obs-studio
+}:
 
-stdenv.mkDerivation {
-  pname = "obs-v4l2sink-unstable";
-  version = "20181012";
+stdenv.mkDerivation rec {
```
Comment (LouisTakePILLz) [inline pkgs/applications/video/obs-studio/v4l2sink.nix:9]: If this should probably be `mkDerivation` (fed by the argset), if `libsForQt5.callPackage` is used, although I'm not sure if `libsForQt5.callPackage` is necessary in the first place.

Thoughts?

Comment (LouisTakePILLz) [reply pkgs/applications/video/obs-studio/v4l2sink.nix:9]: According to #76433, the qt-specific `mkDerivation` will automatically supply `wrapQtAppsHook`.

**Comment by colemickens** [comment]
(Mentioned in IRC, but just for continuity's sake, including here.)

I applied this patch which made the directory structure back to what home-manager expects (and what other plugins tend to follow):

```diff
diff --git a/pkgs/applications/video/obs-studio/v4l2sink.nix b/pkgs/applications/video/obs-studio/v4l2sink.nix
index 29479a33a1b..eddb7928d85 100644
--- a/pkgs/applications/video/obs-studio/v4l2sink.nix
+++ b/pkgs/applications/video/obs-studio/v4l2sink.nix
@@ -35,8 +35,8 @@ stdenv.mkDerivation rec {
   }.${stdenv.targetPlatform.system} or (throw "Unsupported system: ${stdenv.targetPlatform.system}");
 
   postInstall = ''
-    mkdir -p $out/$pluginPath
-    ln -s $out/lib/obs-plugins/v4l2sink.so $out/$pluginPath
+    mkdir -p $out/share/obs/obs-plugins/v4l2sink/$pluginPath
+    ln -s $out/lib/obs-plugins/v4l2sink.so $out/share/obs/obs-plugins/v4l2sink/$pluginPath
   '';
 
   meta = with lib; {

```

**Comment by colemickens** [comment]
Okay, I can confirm - this doesn't crash on Stop anymore. Thanks, looks good.

File: pkgs/applications/video/obs-studio/v4l2sink.nix (hunk 1 above)
Comment (jtojnar) [reply pkgs/applications/video/obs-studio/v4l2sink.nix:21]: All direct dependencies should be listed directly. Adding wrapper just to get some dependencies that maybe match the required dependencies sounds fragile.

File: pkgs/applications/video/obs-studio/v4l2sink.nix (hunk 2 above)
Comment (jtojnar) [reply pkgs/applications/video/obs-studio/v4l2sink.nix:9]: I believe Qt's `mkDerivation` does more harm than good and should not have been introduced. Adding the dependencies (and `wrapQtAppsHook` for programs) should be sufficient. (There is also one minor improvement but it does not outweigh the loss of compositionality.)

Comment (LouisTakePILLz) [reply pkgs/applications/video/obs-studio/v4l2sink.nix:9]: I'll keep it as-is then.

File: pkgs/applications/video/obs-studio/v4l2sink.nix (hunk 1 above)
Comment (LouisTakePILLz) [reply pkgs/applications/video/obs-studio/v4l2sink.nix:21]: Perfect, the PR should be ready for merging then :)

File: pkgs/applications/video/obs-studio/v4l2sink.nix [hunk 3]
```diff
@@ -1,42 +1,49 @@
...
+  pname = "obs-v4l2sink";
+  version = "0.1.0";
```
Comment (jtojnar) [inline pkgs/applications/video/obs-studio/v4l2sink.nix:11]: This seems to have been a downgrade. For software that has been so long without a release, following `HEAD` might be preferable.

File: pkgs/applications/video/obs-studio/v4l2sink.nix (hunk 1 above)
Comment (jtojnar) [inline pkgs/applications/video/obs-studio/v4l2sink.nix:21]: No need to pass `.dev` here, all inputs are always available when passed to `buildInputs`.

File: pkgs/applications/video/obs-studio/v4l2sink.nix [hunk 4]
```diff
@@ -1,42 +1,49 @@
...
   };
 
-  nativeBuildInputs = [ cmake pkgconfig wrapQtAppsHook ];
-  buildInputs = [ obs-studio ];
+  nativeBuildInputs = [ cmake ];
+  buildInputs = [ qt5.qtbase.dev obs-studio ];
 
   patches = [
-    ./0001-find-ObsPluginHelpers.cmake-in-the-obs-src.patch
+    ./fix-v4l2device_close-segfault.patch
```
Comment (jtojnar) [inline pkgs/applications/video/obs-studio/v4l2sink.nix:24]: This should be a compile error, not segfault, if I am reading the patch right.

Moreover could you use `fetchpatch` to obtain `https://github.com/CatxFish/obs-v4l2sink/commit/6604f01796d1b84a95714730ea51a6b8ac0e450b.patch` instead of adding it in-tree? And also add a link to the PR in a comment.

File: pkgs/applications/video/obs-studio/v4l2sink.nix [hunk 5]
```diff
@@ -1,42 +1,49 @@
...
   ];
 
-  cmakeFlags = [
-    "-DLIBOBS_INCLUDE_DIR=${obs-studio}/include/obs"
-    "-DLIBOBS_LIBRARIES=${obs-studio}/lib"
-    "-DCMAKE_CXX_FLAGS=-I${obs-studio.src}/UI/obs-frontend-api"
-    "-DOBS_SRC=${obs-studio.src}"
+  cmakeFlags = with lib; [
+    "-DLIBOBS_INCLUDE_DIR=${obs-studio.src}/libobs"
+    "-DCMAKE_INSTALL_PREFIX=${placeholder "out"}"
```
Comment (jtojnar) [inline pkgs/applications/video/obs-studio/v4l2sink.nix:29]: This should be the default.

File: pkgs/applications/video/obs-studio/v4l2sink.nix [hunk 6]
```diff
@@ -1,42 +1,49 @@
...
-    mkdir -p $out/share/obs/obs-plugins/v4l2sink/bin/64bit
-    cp ./v4l2sink.so $out/share/obs/obs-plugins/v4l2sink/bin/64bit/
+  pluginPath = {
+    i686-linux = "bin/32bit";
+    x86_64-linux = "bin/64bit";
+  }.${stdenv.targetPlatform.system} or (throw "Unsupported system: ${stdenv.targetPlatform.system}");
+
+  postInstall = ''
+    mkdir -p $out/share/obs/obs-plugins/v4l2sink/$pluginPath
+    ln -s $out/lib/obs-plugins/v4l2sink.so $out/share/obs/obs-plugins/v4l2sink/$pluginPath
```
Comment (jtojnar) [inline pkgs/applications/video/obs-studio/v4l2sink.nix:39]: Why is this needed? Could you add a comment?

File: pkgs/applications/video/obs-studio/v4l2sink.nix (hunk 3 above)
Comment (LouisTakePILLz) [reply pkgs/applications/video/obs-studio/v4l2sink.nix:11]: The changes on master are not very important. There's a couple localization fixes and only one build-related change (CatxFish/obs-v4l2sink#1ec3c8ada0e1040d867ce567f177be55cd278378), but it doesn't seem to be critical.

Personally I would rather stay on the release rev.

File: pkgs/applications/video/obs-studio/v4l2sink.nix (hunk 4 above)
Comment (LouisTakePILLz) [reply pkgs/applications/video/obs-studio/v4l2sink.nix:24]: The compile error results in a segfault (if you look at the related issues on the issue tracker).
I've personally had issues with non-deterministic patch/diff output from GitHub (changing hashes). Is it OK if we keep it as a vendored patch since the next version is probably going to include it anyway?

Comment (jtojnar) [reply pkgs/applications/video/obs-studio/v4l2sink.nix:24]: Oh, it leads to UB. :woman_shrugging:

File: pkgs/applications/video/obs-studio/v4l2sink.nix (hunk 6 above)
Comment (LouisTakePILLz) [reply pkgs/applications/video/obs-studio/v4l2sink.nix:39]: obs-studio seems to expects a specific directory structure for plugins, otherwise it won't find it.

File: pkgs/applications/video/obs-studio/v4l2sink.nix (hunk 4 above)
Comment (jtojnar) [reply pkgs/applications/video/obs-studio/v4l2sink.nix:24]: `fetchpatch` is specifically designed to make the patches more deterministic by cleaning the git fluff that changes. Looking at the commit history, I am not sure there will ever be a next release.

File: pkgs/applications/video/obs-studio/v4l2sink.nix (hunk 6 above)
Comment (jtojnar) [reply pkgs/applications/video/obs-studio/v4l2sink.nix:44]: Is that upstream thing or our packaging thing. If the former, why does the project not already do that?

Either way, the comment should be part of the expression.

File: pkgs/applications/video/obs-studio/v4l2sink.nix (hunk 3 above)
Comment (LouisTakePILLz) [reply pkgs/applications/video/obs-studio/v4l2sink.nix]: I reverted the revision change.

Comment (jtojnar) [reply pkgs/applications/video/obs-studio/v4l2sink.nix]: Right, with this diff https://github.com/CatxFish/obs-v4l2sink/compare/0.1.0...master, I would prefer the tag as well. Though the downgrade should be mentioned in the commit message.

File: pkgs/applications/video/obs-studio/v4l2sink.nix (hunk 6 above)
Comment (LouisTakePILLz) [reply pkgs/applications/video/obs-studio/v4l2sink.nix:44]: It's an upstream thing. It looks for plugins in `$XDG_CONFIG_HOME/obs-studio/plugins` and `/usr/share/obs/obs-plugins`

File: pkgs/applications/video/obs-studio/v4l2sink.nix (hunk 3 above)
Comment (LouisTakePILLz) [reply pkgs/applications/video/obs-studio/v4l2sink.nix]: OBS doesn't seem to recognize the plugin since I upgraded to 20.03 (even with the current derivation on master).

Comment (LouisTakePILLz) [reply pkgs/applications/video/obs-studio/v4l2sink.nix:12]: Everything is working again. The downgrade is now documented in the commit message.

File: pkgs/applications/video/obs-studio/v4l2sink.nix [hunk 7]
```diff
@@ -1,42 +1,55 @@
...
+    i686-linux = "bin/32bit";
+    x86_64-linux = "bin/64bit";
+  }.${stdenv.targetPlatform.system} or (throw "Unsupported system: ${stdenv.targetPlatform.system}");
+
+  # obs-studio expects the shared object to be located in bin/32bit or bin/64bit
+  # https://github.com/obsproject/obs-studio/blob/d60c736cb0ec0491013293c8a483d3a6573165cb/libobs/obs-nix.c#L48
+  postInstall = ''
+    mkdir -p $out/share/obs/obs-plugins/v4l2sink/$pluginPath
+    ln -s $out/lib/obs-plugins/v4l2sink.so $out/share/obs/obs-plugins/v4l2sink/$pluginPath
   '';
```
Comment (jtojnar) [inline pkgs/applications/video/obs-studio/v4l2sink.nix:46]: Does the pluginPath need to be in the environment? If not, I would try to reduce the scope:

```suggestion
  # obs-studio expects the shared object to be located in bin/32bit or bin/64bit
  # https://github.com/obsproject/obs-studio/blob/d60c736cb0ec0491013293c8a483d3a6573165cb/libobs/obs-nix.c#L48
  postInstall = let
    pluginPath = {
      i686-linux = "bin/32bit";
      x86_64-linux = "bin/64bit";
    }.${stdenv.targetPlatform.system} or (throw "Unsupported system: ${stdenv.targetPlatform.system}");
  in ''
    mkdir -p $out/share/obs/obs-plugins/v4l2sink/${pluginPath}
    ln -s $out/lib/obs-plugins/v4l2sink.so $out/share/obs/obs-plugins/v4l2sink/${pluginPath}
  '';
```

**Review by jtojnar — APPROVED** [review]
The path shenanigans are weird, looks good otherwise.

**Review by jtojnar — APPROVED** [review]

**Comment by colemickens** [comment]
Pulled latest commits as of right now - still works in my HM-managed setup. LGTM thanks!

**Review by colemickens — APPROVED** [review]

**Comment by jtojnar** [comment]
Thanks.

**Comment by grahamc** [comment]
coming back to this PR for instructions on how to "install" the plugin. Gosh, I would so love a `obs-studioWithPlugins`

**Comment by LouisTakePILLz** [comment]
Yeah that would be a good idea. For now I just have home-manager create a symlink to `${obs-v4l2sink}/share/obs/obs-plugins/v4l2sink` in `~/.config/obs-studio/plugins`.

**Comment by colemickens** [comment]
[From my nixcfg](https://github.com/colemickens/nixcfg/blob/8aac50b1066c4aca861ef8798172fa0ab4c8d52e/home/users/cole/gui.nix#L92): home-manager makes it easy:

```
    home-manager.users.cole = { pkgs, ... }: {
      programs = {
        obs-studio = {
          enable = true;
          plugins = with pkgs; [ obs-wlrobs obs-v4l2sink ];
        };
      };
    };
```

**Comment by Algram** [comment]
@colemickens Hi, thanks for that example! It helped me a lot. Can I ask how you manage the `v4l2loopback` module that is required in order for the obs-v4l2sink plugin to work?

**Comment by davidak** [comment]
@Algram good question. i got the module loaded with 

```
  boot.extraModulePackages = [ pkgs.linuxPackages_latest.v4l2loopback ];
  boot.kernelModules = [ "v4l2loopback" ];
```

and that creates `/dev/video1` but in obs i get the error "format not supported"

Screenshot from 2020-10-11 06-25-03

It would be nice to have documentation for this!
